pkg/distributor/instance_count.go

```diff
@@ -0,0 +1,51 @@
+package distributor
+
+import (
+	"time"
+
+	"github.com/grafana/dskit/ring"
+	"go.uber.org/atomic"
+)
+
+// healthyInstanceDelegate counts the number of healthy instances that are part of the ring
+// and stores the count to the provided atomic integer. Used here to count the number of
+// distributors in the ring to determine how to enforce rate limiting.
+type healthyInstanceDelegate struct {
+	count            *atomic.Uint32
+	heartbeatTimeout time.Duration
+	next             ring.BasicLifecyclerDelegate
+}
+
+func newHealthyInstanceDelegate(count *atomic.Uint32, heartbeatTimeout time.Duration, next ring.BasicLifecyclerDelegate) *healthyInstanceDelegate {
+	return &healthyInstanceDelegate{count: count, heartbeatTimeout: heartbeatTimeout, next: next}
+}
+
+// OnRingInstanceRegister implements the ring.BasicLifecyclerDelegate interface
+func (d *healthyInstanceDelegate) OnRingInstanceRegister(lifecycler *ring.BasicLifecycler, ringDesc ring.Desc, instanceExists bool, instanceID string, instanceDesc ring.InstanceDesc) (ring.InstanceState, ring.Tokens) {
```

Comment:
nit: you could use [struct embedding](https://gobyexample.com/struct-embedding) to only need to define the methods you wish to override.